app/com/arpnetworking/metrics/portal/scheduling/impl/JobScheduler.java

```diff
@@ -0,0 +1,225 @@
+     *
+     * @param <T> The type of result produced by the {@link JobRef}'s job.
+     * @param jobRef The job to intermittently execute.
+     * @param clock The clock the scheduler will use, when it ticks, to determine whether it's time to run the next job(s) yet.
+     * @return A new props to create this actor.
+     */
+    protected static <T> Props props(final JobRef<T> jobRef, final Clock clock) {
+        return Props.create(JobScheduler.class, () -> new JobScheduler<>(jobRef, clock));
+    }
+
+    private JobScheduler(final JobRef<T> jobRef, final Clock clock) {
+        _jobRef = jobRef;
+        _clock = clock;
+        timers().startPeriodicTimer("TICK", Tick.INSTANCE, TICK_INTERVAL);
+    }
+
+    /**
+     * Computes the time we should schedule an extra tick for, if we'll need an extra tick before the next naturally occurring one.
+     *
+     * @param now The current time.
+     * @param timeToAwaken The time we want to wake up right after.
+     * @return The time until we should wake up, if that's before the next tick; else {@code empty}.
+     */
+    protected static Optional<FiniteDuration> timeUntilExtraTick(final Instant now, final Instant timeToAwaken) {
+        final FiniteDuration delta = Duration.fromNanos(ChronoUnit.NANOS.between(now, timeToAwaken));
+        if (delta.lt(TICK_INTERVAL)) {
+            return Optional.of(delta.plus(SLEEP_SLOP));
+        } else {
+            return Optional.empty();
+        }
+    }
+
+    @Override
+    public Receive createReceive() {
+        return receiveBuilder()
+                .match(Tick.class, message -> {
+                    final Optional<Job<T>> maybeJob = _jobRef.get();
+                    if (!maybeJob.isPresent()) {
+                        LOGGER.error()
+                                .setMessage("no such job")
+                                .addData("jobRef", _jobRef)
+                                .log();
+                        return;
+                    }
+                    final Job<T> job = maybeJob.get();
+                    final Optional<Instant> maybeNextRun = job.getSchedule().nextRun(_jobRef.getLastRun());
+                    if (!maybeNextRun.isPresent()) {
+                        getSelf().tell(PoisonPill.getInstance(), getSelf());
+                        return;
+                    }
+                    final Instant nextRun = maybeNextRun.get();
+                    if (nextRun.isAfter(_clock.instant())) {
+                        // It's not time to execute the job yet, but it might be soon.
+                        // Let's ensure we wake up "exactly" when it's time to execute the job,
+                        //   else we might be up to 1tick late executing it.
+                        //   (Well-- "exactly" is a tall order, but we can at least ensure we wake up
+                        //    _very slightly_ late.)
+                        timeUntilExtraTick(_clock.instant(), nextRun).ifPresent(delta -> {
```

Comment:
What happens if `timeUntilExtraTick` returns absent? Then it seems neither scheduling nor job execution happen.

Reply:
Correct! But that's okay, because the actor wakes up every minute anyway. It only schedules an extra tick if it's gonna need to wake up before then.